### PR TITLE
fix: disable remote layer APIs in MAS build

### PR DESCRIPTION
**Pull request by zcbenz**
#### Description of Change

Close https://github.com/electron/electron/issues/20027.

This PR disables following private macOS APIs used in Chromium:

* `CAContext`
* `CALayerHost`

Chromium has 2 ways of rendering on macOS: one is using official `IOSurface` way, and one is `CARemoteLayer` API. The `IOSurface` way is used when software compositing is enabled, and when the remote layer API is not available.

This PR forces using `IOSurface ` even when remote layer API is available, it essentially behaves the same with passing the `--disable-remote-core-animation` command line flag.

More information can be found at https://bugs.chromium.org/p/chromium/issues/detail?id=312462.

* `NSAccessibilityRemoteUIElement`

All related code using this class are disabled. It would affect accessibility when remote layer API is used, but we should be fine since remote layer API is disabled too.

* `NSNextStepFrame`
* `NSThemeFrame`

Chromium overrides these private classes to implement custom frame. All related code are disabled so custom frame would not work.

Since in Electron we only support standard frame and simple frameless window, we are not affected.

* `NSURLFileTypeMappings`

Chromium uses it to guess mime type from extension names, the iOS version of Chromium does not use it. The fallback detection should be enough for us.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fix Electron apps getting rejected to Mac App Store.

### CODE REVIEW

**Comment by zcbenz** [comment]
The original issue should be kept open since this PR does not patch `fileport_makefd` and `fileport_makeport`, and I don't know if it is needed.

(I hope it is not needed since these two are really difficult to get rid of.)

File: patches/chromium/mas_disable_remote_accessibility.patch [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+diff --git a/ui/base/BUILD.gn b/ui/base/BUILD.gn
+index c5d4da4baef2..0101dbb1cf77 100644
+--- a/ui/base/BUILD.gn
++++ b/ui/base/BUILD.gn
+@@ -287,6 +287,13 @@ jumbo_component("base") {
+     "window_open_disposition.h",
+   ]
+ 
++  if (is_mas_build) {
++    sources -= [
```
Comment (deepak1556) [inline patches/chromium/mas_disable_remote_accessibility.patch]: why are we excluding them here ?

**Comment by deepak1556** [comment]
`3-1-x` and `4-2-x` are not supported release lines, hence removed their backport labels

File: patches/chromium/mas_disable_remote_accessibility.patch (hunk 1 above)
Comment (zcbenz) [reply patches/chromium/mas_disable_remote_accessibility.patch]: `remote_accessibility_api.mm` implements `ui::RemoteAccessibility` which uses private APIs, excluding it is simpler than commenting out the whole file.

**Comment by zcbenz** [comment]
> We should look into the possibility of pushing the remote implementation behind a buildflag in upstream to reduce our patch in future. Thoughts ?

It is definitely good idea, it would also keep Chromium maintaining the code path we use.

File: patches/chromium/mas_disable_remote_accessibility.patch (hunk 1 above)
Comment (deepak1556) [reply patches/chromium/mas_disable_remote_accessibility.patch]: But we also seem to be patching them below is that needed ?

Comment (zcbenz) [reply patches/chromium/mas_disable_remote_accessibility.patch]: In theory it is fine leaving `remote_accessibility_api.h` untouched, but by commenting out the declarations defined in `remote_accessibility_api.h`, any other file that uses the private APIs would trigger compilation error early, so we won't miss things when updating/backporting the patch.

**Comment by MarshallOfSound** [comment]
Switching to `IOSurface` will (according to that linked issue) result in a performance hit.  Do we have any idea how to quantify that hit?

**Comment by zcbenz** [comment]
> Switching to `IOSurface` will (according to that linked issue) result in a performance hit. Do we have any idea how to quantify that hit?

I don't know how to measure the performance hit, they did not do any actual measurement neither in the issue.

**Comment by zcbenz** [comment]
Status update: as disabling remote layer APIs might drag down performance significantly, we are currently waiting for responses from Apple to see if it is possible to unblock `CAContext` and `CALayerHost` for us, or whether is there a better way of patching to avoid performance hit.

**Comment by ngehlert** [comment]
any new updates here?

**Comment by ncodeyx** [comment]
Also waiting for update. https://github.com/electron/electron/issues/20027#issuecomment-554785974

**Comment by JCBsystem** [comment]
any news on this ? 
for us with a live app in MAS with live customers this is extremely important

**Comment by soundar78** [comment]
Any idea when this can be merged @MarshallOfSound @nornagon ?

**Comment by gaodeng** [comment]
What can we do to help advance this matter quickly? Or can we release a beta version containing this patch first, so that everyone can easily update the current app on the mac app store

**Comment by JCBsystem** [comment]
new week .. 
please keep us updated

**Comment by gaodeng** [comment]
slack was updated yesterday, does anyone know how they pass the review？I also submitted a new version yesterday, but it was rejected this morning 。

**Comment by JCBsystem** [comment]
> slack was updated yesterday, does anyone know how they pass the review？I also submitted a new version yesterday, but it was rejected this morning 。

@gaodeng what reason did you get rejected for ?
and did you build with the zip or the electron:...path_to_patch_file 
i think path_to_patch_file can have more libs that don't belong i the MAS

**Comment by gaodeng** [comment]
> @gaodeng what reason did you get rejected for ?

Private API Rejection .

> About slack it looks like the use
> "electron": "6.0.10"
> when i look in the package.json
> but they can have used the Zip build hack..

No,I don't think so.
otool -ov with slack's electron framwork 
屏幕快照 2019-12-07 下午4 59 24

Why slack can pass the review？ the answer may be simple  , "All developers are equal, but some developers are more equal than others."

**Comment by JCBsystem** [comment]
Maybe this is a good thing ....
 "we are currently waiting for responses from Apple"  
we can hope apple did a exception for electron :)

**Comment by snokier** [comment]
any news here？

**Comment by gtamas** [comment]
@zcbenz 

Any news? 

Exactly which electron version has this patch? Is there an electron version which is 100% supported  by Apple Store? 

If there isn't, when can we expect such a version to be released?

**Pull request merged**